src/porepy/models/abstract_model.py

```diff
@@ -232,28 +235,42 @@ def _initialize_linear_solver(self) -> None:
         if solver not in ["scipy_sparse", "pypardiso", "umfpack"]:
             raise ValueError(f"Unknown linear solver {solver}")
 
-    def assemble_and_solve_linear_system(self, tol: float) -> np.ndarray:
+    def assemble_linear_system(self):
         """Assemble the linearized system, solve and return the new solution vector.
 
         The linear system is defined by the current state of the model.
 
-        The linear solver is set according to the parameter 'linear_solver' set under
-        initialization of this model.
 
-        Parameters:
-            tol (double): Target tolerance for the linear solver. May be used for
-                inexact approaches.
 
-        Returns:
-            np.ndarray: Solution vector.
+
+
+        Attributes:
+            linear_system is assigned.
 
         """
         t_0 = time.time()
         if self._use_ad:
             A, b = self._eq_manager.assemble()
         else:
             A, b = self.assembler.assemble_matrix_rhs()  # type: ignore
-        t_1 = time.time()
+        self.linear_system = (A, b)
+        logger.debug(f"Assembled linear system in {t_0-time.time():.2e} seconds.")
+
+    def linear_solve(self):
```

Comment:
This is of course a nice default implementation of a linear solver. However, I actually think I would almost always overwrite this method (when considering a time-dependent problem at least). Instead I would try to precompute the LU factorization in a preparing step (e.g. already within `assemble_linear_system`). In `linear_solve`, I would then only call the sparse solver based on the factorization. Even when considering only a single time step, there should be not difference in terms of performance by splitting the sparse solve into the actual components which will be called internally by the solver library. Of course by allowing for different libraries, this introduces a bit more code than just one more line.

If this is no issue, then it is also OK with me.

Reply:
In support of @jwboth comment, I would add a third step like `factorize_lhs()` to precompute LU or LLT factors and keep `assemble_linear_system()` as it is. However, we don't have a separate method for RHS like `assemble_rhs()`, and having a factorization step will be not beneficial because LHS is always computed when you evaluate AD equations. 
In the way I see this PR, the proposed changes are a good starting point and methods like `factorize_lhs()` should be included in a separate pull request dedicated to  implement`assemble_rhs()`.

Reply:
I agree with Omar's conclusion. I see your point, Jakub, but suggest you for now consider factorization part of "assemble" and adjust that method accordingly.

Reply:
Fine for me.